### PR TITLE
Milestone two

### PR DESCRIPTION
**Pull request by wandji20**
This is an app built with the fundamentals of any CRUD application.
A `user` can only see and can create `posts` if logged in.
A `user` can `send`, `accept` or `reject` friend requests.

### CODE REVIEW

**Review by GzimAsani — CHANGES_REQUESTED** [review]
# Status: Changes Required :recycle: 

## To Highlight:
- ### You have done great job so far :muscle: 

## To improve:
- [x] User can see friendships requests ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/pending_request.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
-  When I click once on the 'Invite to friendship' button, in the Users list page, the button doesn't update: it still shows the 'Invite to friendship' text, kindly fix this
- [x] When I reject a friend request it throws me an error `TypeError in FriendshipsController#destroy`
- [x] Please add the number of likes that each posts has
- [ ] Also, please remember to add some integration tests with something like Capybara as it's required by this stage of the project.

### **Please make the required changes and submit another review request when you are ready.**

<hr>
### if you have any question kindly reach me on slack @GzimAsani

### HAPPY CODING :computer:

**Review by descholar-ceo — CHANGES_REQUESTED** [review]
# Status: Changes Required :vertical_traffic_light: :warning: 
## Hello @wandji20 and @helciodev  :handshake: :handshake: :handshake: 

:+1: This is good to make it this far, your project is really nice, congratulations on the good work you have done so far.
:+1: I liked that you have tried to implemented the requested changes from the previous review, keep it up, and your project is really almost done.
#### :warning: However we need to make few changes to mark this project approved, don't feel discouraged because of being asked some changes more and more, that's beacuse this is a learning process, and we are here to help you to become an out-standing developer. See the requested change below :point_down: :point_down: :point_down: 
### Note: All of requested changes which are marked as [Required] are mandatory to be implemented in order to get this project approved, others which are marked as [Optional] are nice to have in your project to make it outstanding

# [Required Changes] :red_circle: 
- [x] Posts timeline includes posts written by logged-in user and their friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from Pathwright)
- [x] When I click once on the 'Invite to friendship' button, in the Users list page, the button doesn't update: it still shows the 'Invite to friendship' text, kindly fix this. Even though we are alredy friends, it's still showing me to `Invite to friendship`
- [x] You have commented codes, let's get rid of them. If you want to read your old odes, you can use git history. There is no need to keep commented codes for that.

# [Optional changes] :yellow_circle: 
- [x] The comment button doesn't show the full text, make sure that is not cut in the middle for the sake of a great User eperience
- [x] The list of `Friends`, `Pending Friendships` , `Sent Friendships should, the displayed names should be links to their profile pages
___
_If anything seems to be unclear to you, reachout and ask anything by commenting on this Pull Request and mention me as [descholar-ceo](https://github.com/descholar-ceo), I will receive a notification and will respond._

[GitHub](https://github.com/descholar-ceo) | [LinkedIn](https://linkedin.com/in/mugirase-emmanuel) | [Twitter](https://twitter.com/descholar3)

_Reviewed by [Mugirase Emmanuel (descholar)](https://github.com/descholar-ceo)_

File: app/views/friendships/_friendship_invite.html.erb
Comment (acushlakoncept) [inline app/views/friendships/_friendship_invite.html.erb]: there should be absolutely NO logic in views

File: app/views/friendships/_friendships_details.html.erb
Comment (acushlakoncept) [inline app/views/friendships/_friendships_details.html.erb]: there should be absolutely NO logic in views

File: app/views/layouts/application.html.erb
Comment (acushlakoncept) [inline app/views/layouts/application.html.erb]: there should be absolutely NO logic in views

File: app/views/posts/_post.html.erb
Comment (acushlakoncept) [inline app/views/posts/_post.html.erb]: there should be absolutely NO logic in views

Comment (acushlakoncept) [inline app/views/posts/_post.html.erb]: there should be absolutely NO logic in views

File: app/views/users/_user.html.erb
Comment (acushlakoncept) [inline app/views/users/_user.html.erb]: there should be absolutely NO logic in views

File: app/models/user.rb
Comment (acushlakoncept) [inline app/models/user.rb]: It’s better to delete **commented piece of codes**. If you need to review a previous version of your code, use git history ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/no_inline_comments.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))
See this [post](https://stackoverflow.com/questions/13130859/inline-comments-in-ruby/13130900#:~:text=No%2C%20Ruby%20does%20not%20have,comment%20out%20the%20one%20row), about why comments are discouraged.

File: spec/models/user_spec.rb
Comment (acushlakoncept) [inline spec/models/user_spec.rb]: It’s better to delete **commented piece of codes**. If you need to review a previous version of your code, use git history ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/no_inline_comments.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

### Hello @helciodev and @wandji20, Great job! 👏  Both of you have done well on this project.

### You've done well so far 👍 and I commend you for that 👏
However, there are still some changes to be made before this project can be approved.

- [x] I noticed there are lots of logic in your views, there should be absolutely NO logic in views. Views are only for displaying information that became available to it from its corresponding controller not generating it.  ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright)
I have gone ahead and made some comments on your codebase:

**Don't know where to start or how to?** See  [Rails View Helpers Best Practices](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4) 

- [ ] You still have commented codes that need to be taken out, see the comments I have made on your codebase:
- [ ] Your tests look good, but you could do more by writing more test coverage for both unit and integration tests. Try using [shoulda-matchers](https://github.com/thoughtbot/shoulda-matchers) and [capybara](https://rubydoc.info/github/teamcapybara/capybara), they will make your testing a lot easier.  You would have something like 
```
  describe 'associations' do
    it { should belong_to(:category).class_name('MenuCategory') }
  end

  describe 'validations' do
    it { should validate_presence_of(:name) }
    it { should validate_uniqueness_of(:name).scoped_to(:category_id) }
  end
```

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla) or reply to this thread, if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Review by enaburekhan — CHANGES_REQUESTED** [review]
# Status: Required changes:full_moon:
**Hi team!
Thank you for implementing the changes requested by the last reviewer:clap::clap:.**

**Major highlights**

- The app is working as expected and login users can request, accept, and reject friendships.
- The codebase is neat and clearly written.
- There are no complex logics in view

However, we just need to make a change to get the project approved:muscle:.

**To improve**
- [x] Currently, the project tests only cover `riendship` and `user` in the model. And `friendship` in the controller. We also need to write tests for `comment`, `post`, and `like` both for model and controller. Both unit and integration test.

Make the above-required changes and when you are ready, request another review.
Kindly drop your questions :point_down: and ping me on slack @ Eric Enaburekhan to respond.
Happy hacking:blush:
Eric Enaburekhan

**Review by ghost — CHANGES_REQUESTED** [review]
# Status: Changes Required 🛑 

### Hello team 👋,

**Great job implementing the requested changes 👏.**

#### Highlights

- **Linters are properly configured.**

### Mandatory requirements

1). **For all your models, you have only written validation tests. Please add association tests, e.g for `belongs_to: user` assocation, and method tests, e.g  `all_friends` method in `user.rb`. Hint: For association tests, you can use [shoulda matchers](http://matchers.shoulda.io/docs/v4.0.1/Shoulda/Matchers/ActiveRecord.html#belong_to-instance_method).**

2). **You have not written tests for `like.rb`.**
3).**Add tests for `like`, `comment` and `user` controllers.**
4). **Add tests for your helpers.**
5). **You have a failing test as shown below.**

rspec


### Happy coding 🥂.

**You can reach me on slack (Rose Wanjohi).**

**Review by enaburekhan — APPROVED** [review]
# Status: Project Approved:heavy_check_mark:
**Hi team!**
**Nice one:thumbsup:. The app is working as expected and the tests cover both positive and negative scenarios. Keep up the good work:clap::clap:.
Move on with the next milestone:walking:.**

Kindly drop your questions :point_down: and ping me on slack @ Eric Enaburekhan to respond
Happy hacking:blush:
Eric Enaburekhan

**Pull request merged**
